File: src/python/fodt/spec/text/paragraph.py
"""FODT spec Paragraph — canonical implementation of text:p.

spec_qname: text:p
spec_fact_ref: SAL-FODT-00003
"""
from __future__ import annotations
from typing import Any, ClassVar

from .span import Span


class Paragraph:
    """Canonical implementation of ODF text:p element."""

    spec_qname: ClassVar[str] = "text:p"
    spec_fact_ref: ClassVar[str] = "SAL-FODT-00003"

    def __init__(self, data: dict[str, Any]):
        self._data = data

    @property
    def kind(self) -> str:
        return self._data.get("kind", "paragraph")

    @property
    def text(self) -> str:
        return self._data.get("text", "")

    @property
    def style_name(self) -> str:
        return self._data.get("style_name", "")

    @property
    def outline_level(self) -> int | None:
        return self._data.get("outline_level")

    @property
    def spans(self) -> list[Span]:
        return [Span(s) for s in self._data.get("spans", [])]

    def set_text(self, text: str) -> None:
        """Mutate the paragraph text in place (TC-PCL-005)."""
        self._data["text"] = str(text)

    def to_dict(self) -> dict[str, Any]:
        return dict(self._data)

    def __repr__(self) -> str:
        return f"Paragraph(kind={self.kind!r}, text={self.text!r})"

File: src/net/fodt/Spec/Text/Paragraph.cs
// FormatFactory.Fodt — Spec.Text.Paragraph — Canonical spec-shaped model class (pilot)
// spec_qname: text:p
// spec_fact_ref: FACT-FODT-003 (ODF 1.3 §5.1.3)
namespace FormatFactory.Fodt.Spec.Text;

/// <summary>
/// Spec-shaped model class for the ODF text:p (paragraph) element.
///
/// ODF 1.3 §5.1.3 — text:p is the canonical paragraph element in ODF text documents.
/// spec_qname: text:p
/// spec_fact_ref: FACT-FODT-003
///
/// This is a canonical class in the Spec/ hierarchy. The facade wrapper is
/// FormatFactory.Fodt.FodtParagraph (in Model/).
/// </summary>
public sealed class Paragraph
{
    /// <summary>The ODF QName for this element. Grounded in ODF 1.3 §5.1.3.</summary>
    public const string SpecQName = "text:p";

    /// <summary>The SAL fact reference for this element (sal-facts-latest.json).</summary>
    public const string SpecFactRef = "FACT-FODT-003";

    /// <summary>The text:style-name attribute value, if present. May be null.</summary>
    public string? StyleName { get; init; }

    /// <summary>The plain-text content of this paragraph element.</summary>
    public string Content { get; init; } = string.Empty;

    /// <summary>Inline spans contained in this paragraph (text:span elements).</summary>
    public IReadOnlyList<string> Spans { get; init; } = [];
}
